vignettes/environments.Rmd

```diff
@@ -0,0 +1,243 @@
+
+```{r}
+source("moo.R")
+moo(5)
+```
+
+We need to do the same thing on the cluster, as by default it would not be found:
+
+```{r}
+id <- task_create_expr(say(moo(5), by = "cow"))
+task_wait(id)
+task_result(id)
+```
+
+The same error as before, but the consequence is different - this is because **our** code is not present.  We update our environment call to add `moo.R`:
+
+```{r}
+hipercow_environment_create(packages = "cowsay", sources = "moo.R")
+```
+
+(note that this replaces the previous definition of `default`).  Now, we can run our task:
+
+```{r}
+id <- task_create_expr(say(moo(5), by = "cow"))
+task_wait(id)
+task_log_show(id)
+```
+
+You can list as many source files as you want.  If you are brave, you could use `dir()` to do:
+
+```
+hipercow_environment_create(
+  sources = dir(glob2rx("R/", pattern = "*.R", full.names = TRUE)))
+```
+
+which would load all `.R` files within the directory `R/` of your hipercow root.  Be careful not to start recursively loading files though.
+
+# Defining globals
+
+A more advanced use of environments is to load large objects so that they are ready when your task runs.  Suppose we have a large object we need to work with, for example a large loaded geojson file or a fit from a model run, and we want to run a function on it.
+
+The obvious thing to do is to write:
+
+```r
+myfn(my_big_data, another_argument)
+```
+
+where 
+
+* `myfn` is our analysis function
+* `my_big_data` is the large object
+* `another_argument` is some other argument we're passing to the call
+
+When we do this, hipercow will save all the inputs to disk, and this could be quite big.  This could be even worse if this is part of a bulk call, where multiple copies may be saved.
+
+This problem can be so bad that we prevent it by default:
+
+```{r, include = FALSE}
+my_big_data <- runif(1e6)
+```
+
+```{r, error = TRUE}
+task_create_expr(length(my_big_data))
+```
+
+So, how do we get this object into our session?  Suppose that `read_big_data()` is the function that we used to read the object in the first place.  We might write a source file `data.R` (the name is not important):
+
+```r
+my_big_data <- read_big_data("bigfile.shp")
+```
+
+```{r, include = FALSE}
+writeLines("my_big_data <- runif(1e6)", "data.R")
+```
+
+and add this to our `sources`, but **also** add `my_big_data` to `globals`.  This tells hipercow that the object `my_big_data` will be globally available and that if it sees it in an expression that you create it should not try and save it:
+
+```{r}
+hipercow_environment_create(sources = "data.R", globals = "my_big_data")
+```
+
+We can now create the task and run it:
+
+```{r}
+id <- task_create_expr(length(my_big_data))
+task_wait(id)
+task_result(id)
+```
+
+If you use this approach, you should be aware of a few things:
+
+* There is no guarantee that the object you see when creating the task is the same as the object loaded by `sources`.  We can add validation if you would find that useful
```

Comment:
this might actually be a lie and is controlled by `hipercow.validate_globals` https://mrc-ide.github.io/hipercow/articles/details.html#hipercow-validate_globals

Reply:
ok, updated now